### PR TITLE
Add function to get all hud elements

### PR DESCRIPTION
**Pull request by cx384**
- Currently it is not possible to get all HUD element IDs. If you would try to enumerate them, it is impossible to know when to stop. You can't just stop after an ID is empty, since they are not consecutive if an element gets removed.
- This PR adds a `player:hud_get_all()` function which returns a table indexed by all present HUD ID's and containing the corresponding HUD element definition structures.
- Like the `get_lists()` function for inventories, it allows bulk access and makes it easy to iterate through all HUD elements.
- This PR adds the new function to the Lua API as well as to the client Lua API.

#### Does it resolve any reported issue?
No.
#### Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
No.

## To do
Ready for Review.

## How to test
- Start devtest
- Execute the `/hudprint` chat command
- To test some more, change the HUD with for example:
 ```
local def = {
	hud_elem_type = "text",
	name = "test",
	position = {x = 0.5, y = 0.5},
	text = "test"
}
player:hud_add(def)
local rid = player:hud_add(def)
player:hud_add(def)
player:hud_remove(rid)

 ```
and use `/hudprint` again or `print(dump(player:hud_get_all()))` 

To check whether the ids are correct
 ```
for id, elem in pairs(player:hud_get_all()) do
	if dump(elem) ~= dump(player:hud_get(id)) then
		print("Wrong ids")
	end
end
 ```

### CODE REVIEW

File: src/script/lua_api/l_localplayer.cpp [hunk 1]
```diff
@@ -422,6 +422,25 @@ int LuaLocalPlayer::l_hud_get(lua_State *L)
...
+	if (player == nullptr)
+		return 0;
+
+	lua_newtable(L);
+	u32 max = player->getHudIdMax();
+	for (u32 id = 0; id < max; id++) {
+		HudElement *elem = player->getHud(id);
+		if (elem != nullptr) {
+			push_hud_element(L, elem);
+			lua_rawseti(L, -2, id+1);
```
[the remark this answers is not in the document]
Comment (cx384) [reply src/script/lua_api/l_localplayer.cpp]: Oh thank you. I did this out of reflex because Lua tables are 1-indexed.
I should have tested whether the indices are correct.
 
Fixed

**Comment by appgurueu** [comment]
> Currently it is not possible to get all HUD element IDs. If you would try to enumerate them, it is impossible to know when to stop. You can't just stop after an ID is empty, since they are not consecutive if an element gets removed.

I think it could still be possible: Add HUD elements until you get two consecutive IDs. That gives you an upper bound up until which you could enumerate HUD elements ;)

On a more serious note, what's the use case for this? It seems to me so far that forcing modders to keep track of HUD IDs themselves is the better solution.

**Comment by cx384** [comment]
> I think it could still be possible: Add HUD elements until you get two consecutive IDs. That gives you an upper bound up until which you could enumerate HUD elements ;)

There could always be a gap of more than two IDs, if they get removed, so I don't see how this could be a solution, but maybe I didn't understand you correctly. Also currently removed IDs get reused, so you don't get much information from adding new elements. (Maybe HUD IDs shouldn't be reused, but unique anyways.)

In practice you could just assume that there are at most 1000 or so elements, since it is very unlikely that there are more, but this would just be inefficient.



> On a more serious note, what's the use case for this? It seems to me so far that forcing modders to keep track of HUD IDs themselves is the better solution.

Mods should still keep track of their own HUD IDs, but most of the time they don't expose them for others to see. So from another mod you can't know for sure which elements exist.

Some use cases:
- Change the z_index of all existing elements.
- Overwrite all elements of a certain kind.
- Find a free HUD position where there isn't already another element .

**Comment by appgurueu** [comment]
> > On a more serious note, what's the use case for this? It seems to me so far that forcing modders to keep track of HUD IDs themselves is the better solution.
> 
> Mods should still keep track of their own HUD IDs for identification, but most of the time they don't expose them for others to see. So from another mod you can't know for sure which elements exist.

Yes, and I think this is a feature: You are forced to use the APIs of other mods (or your own APIs / tables) to manage HUD elements. This helps prevent mod incompatibilities since it makes it harder to "pull the rug" from under other mods.

Frankly, I'm worried that such an API would be used mostly irresponsible. The only "responsible use" I can imagine is one regarding HUD elements you control yourself, which is already what modders do given the current API(s), and which is probably cleaner (and more efficient) to keep this way. I'm not really opposed to adding such a feature, but I believe it should probably at least get a warning in the docs, considering e.g. your use cases, all of which have high potential to "pull the rug" from under other mods:

* "Change the z_index of all existing elements" - suddenly the mods' assumptions about z-index conventions are broken. (I also believe you should almost never change z-indices of existing elements, esp. of other mods - rather, you should pick your own z-indices of the new elements you're adding appropriately.)
* "Overwrite all elements of a certain kind" - this is extremely likely to introduce incompatibilities.
* "Find a free HUD position where there isn't already another element" - this seems like a valid use case (though something like a "layouting system" would probably be better for this).

Effectively, whenever you're messing with "foreign" HUD elements, you're likely to break stuff. Using this API to get a hold of HUD IDs to modify foreign elements should thus be discouraged. Furthermore, using this API to let Minetest manage your own elements should be discouraged as well: Managing them yourself is cleaner and more efficient.

**TL;DR:** You're right, this feature is useful, as esp. your last use case shows. I'm worried that modders will abuse this though, so I'd recommend that the docs discourage (1) messing with foreign HUD elements; (2) foregoing keeping track of their HUD elements themselves.

File: games/devtest/mods/testhud/init.lua [hunk 2]
```diff
@@ -209,3 +209,23 @@ minetest.register_on_leaveplayer(function(player)
...
 end)
+
+minetest.register_chatcommand("hudprint", {
+	description = "Writes all used Lua HUD elements into chat.",
+	func = function(name, params)
+		local player = minetest.get_player_by_name(name)
+		if not player then
+			return false, "No player."
+		end
+
```
Comment (SmallJoker) [inline games/devtest/mods/testhud/init.lua:220]: Tailing whitespace
*How did I notice that?*
```
.git/rebase-apply/patch:55: trailing whitespace.
		
.git/rebase-apply/patch:63: trailing whitespace.
		
.git/rebase-apply/patch:195: trailing whitespace.
	
```
in `git am -3 patchfile`

File: games/devtest/mods/testhud/init.lua [hunk 3]
```diff
@@ -209,3 +209,23 @@ minetest.register_on_leaveplayer(function(player)
...
+		end
+		
+		local s = "HUD elements:"
+		for k, elem in pairs(player:hud_get_elements()) do
+			local ename = dump(elem.name)
+			local etype = dump(elem.type)
+			local epos = "{x="..elem.position.x..", y="..elem.position.y.."}"
+			s = s.."\n["..k.."]  type = "..etype.." | name = "..ename.." | pos = ".. epos
+		end
+
```
Comment (SmallJoker) [inline games/devtest/mods/testhud/init.lua:228]: Tailing whitespace

File: src/script/lua_api/l_object.h [hunk 4]
```diff
@@ -287,6 +287,9 @@ class ObjectRef : public ModApiBase {
 
 	// hud_get(self, id)
 	static int l_hud_get(lua_State *L);
+
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.h:290]: Tailing whitespace

File: src/script/lua_api/l_localplayer.cpp [hunk 5]
```diff
@@ -422,6 +422,25 @@ int LuaLocalPlayer::l_hud_get(lua_State *L)
...
+int LuaLocalPlayer::l_hud_get_elements(lua_State *L)
+{
+	LocalPlayer *player = getobject(L, 1);
+	if (player == nullptr)
+		return 0;
+
+	lua_newtable(L);
+	u32 max = player->getHudIdMax();
+	for (u32 id = 0; id < max; id++) {
+		HudElement *elem = player->getHud(id);
```
Comment (SmallJoker) [inline src/script/lua_api/l_localplayer.cpp:435]: Please get the entire list of active HUD IDs from the server. This loop is very inefficient.

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -1596,6 +1596,27 @@ int ObjectRef::l_hud_get(lua_State *L)
...
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkObject<ObjectRef>(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == nullptr)
+		return 0;
+
+	lua_newtable(L);
+	u32 max = player->getHudIdMax();
+	for (u32 id = 0; id < max; id++) {
+		HudElement *elem = player->getHud(id);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp:1611]: Inefficient loop here as well

**Review by SmallJoker — CHANGES_REQUESTED** [review]

File: games/devtest/mods/testhud/init.lua (hunk 2 above)
Comment (cx384) [reply games/devtest/mods/testhud/init.lua]: It's fixed. Sorry, for the incontinence. Why isn't this checked by the CI/CD pipeline? o.o
My editor now automatically removes them.

File: games/devtest/mods/testhud/init.lua (hunk 3 above)
Comment (cx384) [reply games/devtest/mods/testhud/init.lua]: fixed

File: src/script/lua_api/l_object.h (hunk 4 above)
Comment (cx384) [reply src/script/lua_api/l_object.h]: fixed

File: src/script/lua_api/l_object.cpp (hunk 6 above)
Comment (cx384) [reply src/script/lua_api/l_object.cpp]: fixed

Comment (cx384) [reply src/script/lua_api/l_object.cpp]: fixed

File: src/script/lua_api/l_localplayer.cpp (hunk 5 above)
Comment (cx384) [reply src/script/lua_api/l_localplayer.cpp]: fixed

Comment (cx384) [reply src/script/lua_api/l_localplayer.cpp]: I did this, because I was unsure about the Mutex in the Player class.
Now it use a function pointer, hopefully it is faster now.

File: src/script/lua_api/l_localplayer.cpp [hunk 7]
```diff
@@ -430,14 +430,15 @@ int LuaLocalPlayer::l_hud_get_elements(lua_State *L)
...
 
 	lua_newtable(L);
-	u32 max = player->getHudIdMax();
-	for (u32 id = 0; id < max; id++) {
-		HudElement *elem = player->getHud(id);
-		if (elem != nullptr) {
-			push_hud_element(L, elem);
-			lua_rawseti(L, -2, id);
+	player->hudApply([&](const std::vector<HudElement*>& hud) {
+		for (std::size_t id = 0; id < hud.size(); ++id) {
```
Comment (appgurueu) [inline src/script/lua_api/l_localplayer.cpp:434]: Couldn't you do something like `(for auto elem : hud)` here (same below)?

File: src/script/lua_api/l_localplayer.cpp [hunk 8]
```diff
@@ -430,14 +430,15 @@ int LuaLocalPlayer::l_hud_get_elements(lua_State *L)
...
+			HudElement *elem = hud[id];
+			if (elem != nullptr) {
```
Comment (appgurueu) [inline src/script/lua_api/l_localplayer.cpp:436]: Can the elements still be nullptrs if you're iterating over the vector of active elements?

File: src/script/lua_api/l_localplayer.cpp (hunk 7 above)
Comment (cx384) [reply src/script/lua_api/l_localplayer.cpp:434]: No, I need to get the id.

File: src/script/lua_api/l_localplayer.cpp (hunk 8 above)
Comment (cx384) [reply src/script/lua_api/l_localplayer.cpp:436]: Yes, they are nullptrs when they get removed.
https://github.com/minetest/minetest/blob/d1a55e9ca4e66d8ba2d799569de7a6b48bc15f23/src/player.cpp#L142-L152

This is the main reason why we need the `player:hud_get_elements()` feature.

File: src/script/lua_api/l_localplayer.cpp (hunk 5 above)
Comment (SmallJoker) [reply src/script/lua_api/l_localplayer.cpp]: Now it looks uglier but is definitely faster.

File: doc/client_lua_api.md [hunk 9]
```diff
@@ -1152,6 +1152,10 @@ Methods:
     * See [`HUD definition`](#hud-definition-hud_add-hud_get)
 * `hud_get(id)`
     * returns the [`definition`](#hud-definition-hud_add-hud_get) of the HUD with that ID number or `nil`, if non-existent.
+* `hud_get_elements()`:
+    * returns a table indexed by all present HUD IDs and containing corresponding HUD element definition structures.
+    * A mod should keep track of its introduced IDs and only use this to access foreign elements.
+    * It is discourage to change foreign HUD elements.
```
Comment (SmallJoker) [inline doc/client_lua_api.md:1158]:
```suggestion
    * It is discouraged to change foreign HUD elements.
```

Same for lua_api.md

File: doc/client_lua_api.md [hunk 10]
```diff
@@ -1152,6 +1152,10 @@ Methods:
     * See [`HUD definition`](#hud-definition-hud_add-hud_get)
 * `hud_get(id)`
     * returns the [`definition`](#hud-definition-hud_add-hud_get) of the HUD with that ID number or `nil`, if non-existent.
+* `hud_get_elements()`:
+    * returns a table indexed by all present HUD IDs and containing corresponding HUD element definition structures.
```
Comment (SmallJoker) [inline doc/client_lua_api.md:1156]:
```suggestion
    * Returns a table in the form `{ [id] = HUD definition, [id] = ... }`.
```
Shorter and clearly describes what the key and values are, respectively.

File: src/player.cpp [hunk 11]
```diff
@@ -139,6 +139,12 @@ HudElement* Player::getHud(u32 id)
 	return NULL;
 }
 
+void Player::hudApply(std::function<void(const std::vector<HudElement*>&)> f)
+{
+	MutexAutoLock lock(m_mutex);
+	f(hud);
```
Comment (SmallJoker) [inline src/player.cpp:145]: This mutex locking issue could be solved without a callback function: https://gist.github.com/SmallJoker/f75cc0e029e1ee989d59e88eab1138cc

I do not know whether other devs like it, thus I am leaving it here as an idea for feedback.

File: doc/lua_api.md [hunk 12]
```diff
@@ -7772,6 +7772,10 @@ child will follow movement and rotation of that bone.
     * `stat` supports the same keys as in the hud definition table except for
       `"hud_elem_type"`.
 * `hud_get(id)`: gets the HUD element definition structure of the specified ID
+* `hud_get_elements()`:
```
Comment (grorp) [inline doc/lua_api.md:7775]: Since `hud_get` is not called `hud_get_element`, `hud_remove` is not called `hud_remove_element` etc., it is inconsistent naming that the name of `hud_get_elements` contains the word "elements".

`hud_get_all` would be more consistent.

File: doc/client_lua_api.md (hunk 10 above)
Comment (cx384) [reply doc/client_lua_api.md]: Changed

File: doc/client_lua_api.md (hunk 9 above)
Comment (cx384) [reply doc/client_lua_api.md]: Fixed

File: doc/lua_api.md (hunk 12 above)
Comment (cx384) [reply doc/lua_api.md:7775]: The HUD does not only consist of elements so I thought `hud_get_elements` would be more descriptive, but for consistency `hud_get_all` is definitely better.

It is changed now.

File: src/player.cpp (hunk 11 above)
Comment (cx384) [reply src/player.cpp:145]: Imo, the callback is fine.
Carrying a PtrLock everywhere is also not so nice.
Alternatively the mutex could be made available outside of the Player class, but this would just scatter things.

However, if anyone has a better idea that is supported by the core devs please make a suggestion.

**Review by appgurueu — APPROVED** [review]
Tested, works. Code is fine too. (Not sure we want to nothing vs. `nil`, but that's a nitpick; also not duplicating HUD code would be ideal, but is out of scope for this PR to resolve.)

**Review by SmallJoker — APPROVED** [review]
Works on my PC.
